packages/datadog-plugin-openai/src/tracing.js

```diff
@@ -0,0 +1,1028 @@
+        retrieveModelRequestExtraction(tags, payload)
+        break
+
+      case 'listFineTuneEvents':
+      case 'fine_tuning.jobs.listEvents':
+      case 'fine-tune.listEvents':
+      case 'retrieveFineTune':
+      case 'fine_tuning.jobs.retrieve':
+      case 'fine-tune.retrieve':
+      case 'deleteModel':
+      case 'models.del':
+      case 'cancelFineTune':
+      case 'fine_tuning.jobs.cancel':
+      case 'fine-tune.cancel':
+        commonLookupFineTuneRequestExtraction(tags, payload)
+        break
+
+      case 'createEdit':
+      case 'edits.create':
+        createEditRequestExtraction(tags, payload, openaiStore)
+        break
+    }
+
+    span.addTags(tags)
+
+    ctx.currentStore = { ...store, span, openai: openaiStore }
+
+    return ctx.currentStore
+  }
+
+  asyncEnd (ctx) {
+    const { result } = ctx
+    const store = ctx.currentStore
+
+    const span = store?.span
+    if (!span) return
+
+    const error = !!span.context()._tags.error
+
+    let headers, body, method, path
+    if (!error) {
+      headers = result.headers
+      body = result.data
+      method = result.request.method
+      path = result.request.path
+    }
+
+    if (!error && headers?.constructor.name === 'Headers') {
+      headers = Object.fromEntries(headers)
+    }
+    const methodName = span._spanContext._tags['resource.name']
+
+    body = coerceResponseBody(body, methodName)
+
+    const openaiStore = store.openai
+
+    if (!error && (path?.startsWith('https://') || path?.startsWith('http://'))) {
+      // basic checking for if the path was set as a full URL
+      // not using a full regex as it will likely be "https://api.openai.com/..."
+      path = new URL(path).pathname
+    }
+    const endpoint = lookupOperationEndpoint(methodName, path)
+
+    const tags = error
+      ? {}
+      : {
+          'openai.request.endpoint': endpoint,
+          'openai.request.method': method.toUpperCase(),
+
+          'openai.organization.id': body.organization_id, // only available in fine-tunes endpoints
+          'openai.organization.name': headers['openai-organization'],
+
+          'openai.response.model': headers['openai-model'] || body.model, // specific model, often undefined
+          'openai.response.id': body.id, // common creation value, numeric epoch
+          'openai.response.deleted': body.deleted, // common boolean field in delete responses
+
+          // The OpenAI API appears to use both created and created_at in different places
+          // Here we're conciously choosing to surface this inconsistency instead of normalizing
+          'openai.response.created': body.created,
+          'openai.response.created_at': body.created_at
+        }
+
+    responseDataExtractionByMethod(methodName, tags, body, openaiStore)
+    span.addTags(tags)
+
+    this.sendLog(methodName, span, tags, openaiStore, error)
+    this.sendMetrics(headers, body, endpoint, span._duration, error, tags)
+  }
+
+  finish (ctx) {
```

Comment:
APM reviewers: This is the only function I added from the original implementation of this plugin. Relates to the point raised in the PR description.